homeassistant/config.py

```diff
@@ -663,6 +666,27 @@ async def async_process_ha_core_config(hass: HomeAssistant, config: dict) -> Non
     if CONF_UNIT_SYSTEM in config:
         hac.units = get_unit_system(config[CONF_UNIT_SYSTEM])
 
+    async def load_translations(_event: Event) -> None:
+        language = hass.config.language
+        _LOGGER.debug(f"Loading translations for language: {language}")
+        await async_load_state_translations_to_cache(hass, language)
+
+    async def load_translations_for_component(event: Event) -> None:
+        component = event.data.get("component")
+        if component is None or "." in str(component):
+            _LOGGER.debug(
+                f"Skipping loading translations for an invalid component: {component}"
+            )
+            return
+        language = hass.config.language
+        _LOGGER.debug(
+            f"Loading translations for language: {hass.config.language} and component: {component}"
+        )
+        await async_load_state_translations_to_cache(hass, language, [component])
+
+    hass.bus.async_listen(EVENT_COMPONENT_LOADED, load_translations_for_component)
+    hass.bus.async_listen(EVENT_CORE_CONFIG_UPDATE, load_translations)
```

Comment:
This is fired twice at startup, with no event data from this method:
https://github.com/home-assistant/core/blob/c433b251fa125789a2b8a68e67014d0eeb7bf6e0/homeassistant/core.py#L379-L380

It's also fired when the configuration is changed, with the configuration change included in the event:
https://github.com/home-assistant/core/blob/c433b251fa125789a2b8a68e67014d0eeb7bf6e0/homeassistant/core.py#L2226-L2227

This means we can improve `load_translations` to only check the event data and return nothing if the language has not been changed.

Reply:
It can be changed the following way to handle situations when it's fired with no data during startup:
```python
    async def load_translations(event: Event) -> None:
        if "language" in event.data:
            language = hass.config.language
            _LOGGER.debug(f"Loading translations for language: {language}")
            await async_load_state_translations_to_cache(hass, language)
```

Is there a way to detect if specifically language has been changed in the config?

Reply:
Yes, if the event has data, the data has the changes: https://github.com/home-assistant/core/blob/b5df7debfb1d85323a067510e12bb511610e780f/homeassistant/core.py#L2534

Reply:
It seems that if config change is made on `/config/general` page then event data always contain all fields from the form, not only changed ones: 

```
## FIRED DURING HA START
2024-02-03 07:43:59.748 DEBUG (MainThread) [homeassistant.helpers.translation] EVENT_CORE_CONFIG_UPDATE: {'event_type': 'core_config_updated', 'data': {}, 'origin': 'LOCAL', 'time_fired': '2024-02-03T06:43:59.747253+00:00', 'context': {'id': '01HNPVBBM3RC6QJ4AF1DGBGWDB', 'parent_id': None, 'user_id': None}}
2024-02-03 07:43:59.762 DEBUG (MainThread) [homeassistant.helpers.translation] EVENT_CORE_CONFIG_UPDATE: {'event_type': 'core_config_updated', 'data': {}, 'origin': 'LOCAL', 'time_fired': '2024-02-03T06:43:59.761235+00:00', 'context': {'id': '01HNPVBBMHJ23MXD5ZGHRS1F1G', 'parent_id': None, 'user_id': None}}

# FIRED AFTER ELEVATION CHANGE
2024-02-03 07:44:32.221 DEBUG (MainThread) [homeassistant.helpers.translation] EVENT_CORE_CONFIG_UPDATE: {'event_type': 'core_config_updated', 'data': {'currency': 'PLN', 'elevation': 1, 'unit_system': 'metric', 'time_zone': 'Europe/Warsaw', 'location_name': 'Dom', 'language': 'pl', 'country': 'PL'}, 'origin': 'LOCAL', 'time_fired': '2024-02-03T06:44:32.220893+00:00', 'context': {'id': '01HNPVCBAWM293VHFCDNVM1YGF', 'parent_id': None, 'user_id': None}}

# FIRED AFTER LANGUAGE CHANGE
2024-02-03 07:44:52.709 DEBUG (MainThread) [homeassistant.helpers.translation] EVENT_CORE_CONFIG_UPDATE: {'event_type': 'core_config_updated', 'data': {'currency': 'PLN', 'elevation': 1, 'unit_system': 'metric', 'time_zone': 'Europe/Warsaw', 'location_name': 'Dom', 'language': 'en', 'country': 'PL'}, 'origin': 'LOCAL', 'time_fired': '2024-02-03T06:44:52.709609+00:00', 'context': {'id': '01HNPVCZB50V15Q2EMGQSWEE77', 'parent_id': None, 'user_id': None}}

# FIRED AFTER COUNTRY CHANGE
2024-02-03 07:45:36.475 DEBUG (MainThread) [homeassistant.helpers.translation] EVENT_CORE_CONFIG_UPDATE: {'event_type': 'core_config_updated', 'data': {'currency': 'PLN', 'elevation': 1, 'unit_system': 'metric', 'time_zone': 'Europe/Warsaw', 'location_name': 'Dom', 'language': 'en', 'country': 'GB'}, 'origin': 'LOCAL', 'time_fired': '2024-02-03T06:45:36.474759+00:00', 'context': {'id': '01HNPVEA2TEZ95SP9MJRGS0F5A', 'parent_id': None, 'user_id': None}}
```